protocols/relay/src/lib.rs

```diff
@@ -23,9 +23,56 @@
 
 #![cfg_attr(docsrs, feature(doc_cfg, doc_auto_cfg))]
 
+mod behaviour;
+mod copy_future;
+mod priv_client;
+mod protocol;
 pub mod v2;
 
+#[allow(clippy::derive_partial_eq_without_eq)]
+mod message_proto {
+    include!(concat!(env!("OUT_DIR"), "/message_v2.pb.rs"));
+}
+
+pub use behaviour::{Behaviour, CircuitId, Config, Event};
+pub use protocol::{HOP_PROTOCOL_NAME, STOP_PROTOCOL_NAME};
+
+/// Types related to the relay protocol inbound.
+pub mod inbound {
+    pub mod hop {
+        pub use crate::protocol::inbound_hop::FatalUpgradeError;
+    }
+    pub mod stop {
+        pub use crate::protocol::inbound_stop::FatalUpgradeError;
+    }
+}
+
+/// Types related to the relay protocol outbound.
+pub mod outbound {
+    pub mod hop {
+        pub use crate::protocol::outbound_hop::FatalUpgradeError;
+    }
+    pub mod stop {
+        pub use crate::protocol::outbound_stop::FatalUpgradeError;
+    }
+}
+
+/// Everything related to the relay protocol from a client's perspective.
+pub mod client {
+    pub use crate::priv_client::{new, Behaviour, Event, RelayedConnection, Transport};
```

Comment:
Uh I just had a closer look at `RelayedConnection` and saw that this one is a public enum with fairly complex variants! (An absolute semver hazard!) I think we should encapsulate that one into an opaque type that only exposes `AsyncRead` and `AsyncWrite` as its interface, i.e.:

- Rename the type to `RelayedConnectionState` & make it private
- Create a struct `RelayedConnection` (or maybe just `Connection` to stick with the naming convention?)
- Move the `AsyncRead` and `AsyncWrite` implementations from the enum to the struct

This will allow us in the future to change the layout of this enum without making a breaking change.

This is a breaking change now though but I'll happily take that for the benefit of sealing the API.

Reply:
This could happen in a separate PR though, perhaps in this one, we just rename it to `Connection` using a type-alias?

Reply:
> Yeah makes sense, created https://github.com/libp2p/rust-libp2p/issues/3255 Btw, there are more `Relayed` prefixed types, should we also address them?

Where it makes sense, sure. I think there is a `RelayedMultiaddr`, not sure if we can rename that to something better. As long as it is private, I don't really mind :)

Reply:
yeah apart from that only `RelayedDial` was left, renamed it to `Dial`.